### DISCUSSION TITLE
Can I use JetStream KV as cache?

### DISCUSSION THREAD

**Discussion by tzq0301** [nats-io/nats.go#1507] [Q&A] [answered]
I tried to use JetStream KV as cache, but it seems to get worse than even no cache. And Redis goes well in the case. 

I wonder if my test case is not good (not suitable for the cache-case). 

The test code is [nats-demo/cache](https://github.com/tzq0301/nats-demo/tree/master/cache)

**Comment by ripienaar** [comment]
Try setting the KV to memory storage.

**Reply by tzq0301** [reply]
Do you mean in-memory cache? But what I need is distributed cache like Redis.

**Reply by ripienaar** [reply]
Redis stores its data in memory. So can nats and it would be faster than going to disk all the time when not needed. 

For like for like comparison you should set KV to memory

**Reply by tzq0301** [reply]
Oh I find the type argument in `jetstream.KeyValueConfig`, and I set `Storage: jetstream.MemoryStorage` in `func BenchmarkNatsCache`. However, it stays slower than no-cache. Does my code uses the KV in a wrong way?

https://github.com/tzq0301/nats-demo/blob/2611e5eac7e7ca43a24619284dc5552df9b588c6/cache/post_test.go#L85

**Reply by ripienaar** [reply]
I don’t really think a cache is what our KV is going to be the best at, it’s not really been a goal we are more into distributed cluster scale stuff with replicated data etc. but it shouldn’t be bad

Regarding rhe test, be aware that MySQL when doing queries like this is incredibly fast since the data is almost all in RAM cached already and it’s optimised for that with its own caches so you aren’t really testing realistic use cases.

**Reply by tzq0301** [reply]
1. I don't understand that 'the data is almost all in RAM cached already'. MySQL is a disk-based storage and cache of MySQL has been deprecated since MySQL 8.
2. I don't understand that 'so you aren’t really testing realistic use cases'. In the benchmark, Redis do fasten the query and you can `go test -bench=` or just see the README.
3. I just try to use NATS as cache because someone in the slack (jetstream) say that it's fast enough.

<img width="389" alt="image" src="https://github.com/nats-io/nats.go/assets/58631544/6ec57949-a407-422e-bc4a-c7b2444ed097">

If my test code is poor, looking forward to a guide.

**Reply by ripienaar** [reply]
While the query cache is indeed gone there are many other layers of caching happening. Block level caching in the database and lots in the OS. 

From your OS and up with that little data you can see with your OS I/O monitoring tools that you would almost never be reading data from disk. 

Better test is much more data - much more, more intensive queries reads in more random ways. With such a simple test the return on investment of the cache would be minimal. 

Again Nats KV isn’t optimised as a very fast cache imo though

**Comment by derekcollison** [comment, marked as answer]
We can be on par with Redis when there is a real network.

You can also use KV watchers to have a subset of the KV space in your app, and kept up to date for even faster response times.

I disagree with R.I., no reason it can not be very fast IMO.

Should make sure direct get acces is on with your KV etc.

**Comment by jnmoyne** [comment]
To add a bit more concrete details to the answer above. How you can do a simple NATS and Redis latency comparison using the KV get() operation: 

NATS setup is `nats-server -js` running locally on the machine and then using the `nats bench` command:
Populate 1M keys `nats bench test --kv --pub 1 --msgs 1000000 --storage memory` and then read them `nats bench test --kv --sub 1 --msgs 1000000 --storage memory --no-progress` 

Redis setup is just start `redis-server` running locally and then using the `redis-benchmark` command:
Populate 1M keys `redis-benchmark -d 128 -c 1 -r 1000000 -n 1000000 -t set` then read them `redis-benchmark -d 128 -c 1 -r 1000000 -n 1000000 -t get`

On a Mac Studio M1-Ultra the NATS benchmark reports almost exactly 24,000 synchronous operations per second meaning an average latency of 41.666 micro-seconds per get. The redis benchmark reports an average of 18 micro-seconds. 

These test are between processes on the same host over the loopback interface, in real-life deployments however client applications and NATS or Redis servers are going to talk to each other over the network, and that latency due to the network is going to be a number of times larger than the latency of NATS/Redis as measured above (e.g. in the hundreds of micro-seconds over a fast LAN) such that the 20 or so micro-seconds of difference in latency between NATS and Redis becomes relatively insignificant.

**Reply by tzq0301** [reply]
Thanks for your answer. Learn a lot!

**Reply by nickchomey** [reply]
If the nats nodes are on the same machine as the application that is using it as a kv cache (ie no network latency), is it safe to assume that the read latency should be close to the 41 and 18 microseconds that you got in the benchmark? This is how I intend to use NATS with my application

**Reply by jnmoyne** [reply]
You can look at the `nats bench` code, it's just using the KV's get() call from the standard nats.go client library, no magic involved. This numbers I got on my M1-Ultra machine: your mileage may vary. The whole point of `nats bench` is that you can easily install it on your target infrastructure and run it there to get an actual performance number.

**Comment by magicprinc** [comment]
NATS KV on my local PC (SSD, 32CPU, Windows 10)

I need very fast "cache" for a very large dataset.
The dataset is in MSSQL. But you random key-value access to MSSQL is too slow for this task.
RAM is not enough to keep the whole dataset.


`nats bench kv put --size="128" --msgs 1000000 --storage file`
1m3s → Pub stats: 15,656 msgs/sec ~ 1.91 MB/sec

`nats bench kv get --size="128" --msgs 1000000`
59s → Sub stats: 16,720 msgs/sec ~ 2.04 MB/sec

This is not fast enough 😱
MSSQL (batches) gives me 42_000 op/sec for "put" 😭

I think about self-made KV: NATS Core ASYNC + Facebook RocksDB Java (local key-value DB)
(or probably NATS JetStream + RocksDB for consistency)

Current (quick and dirty) benchmark shows:
put op/s: 235_938 (similar to NATS publishAsync 214_041)
get op/s: 547_255

**Reply by jnmoyne** [reply]
`nats bench kv put` does synchronous puts, so the throughput is pretty low because it's limited by having to do one RTT for each put. Try parallelizing a bit (or if you were writing an application, you could make asynchronous calls to do your KV puts (i.e. like batches)).
e.g. `nats bench kv put --msgs 1000000 --clients 10` gives me over 43k msgs/s, 25 clients over 55k msgs/s, etc...

The same way `nats bench kv get` also does synchronous get operations, so the same was as the put with just one client it's the RTT that drives the throughput you get (which is why you get around the same number of messages/s in your put and get benchmarks). What you need to look at is the latency of each operation so in your case 1/16720=0,00005980861244 seconds latency (or almost 60 microseconds). The same way as for puts if you want a lot of throughput you have to do a 'batch' of asynchronous get requests, or increase the number of clients. In my case with 1 client I get 18k/s, 10 clients 54k/s, 25 clients 66k/s.

`nats bench kv` doesn't do async batches for put/get, but if you download the current top of main for `nats` it now has a command to benchmark batched direct gets `nats bench js get` and you can get an idea of how much faster it can go (e.g. hundreds of thousands/s or more). And also take note that those get operations as 'direct get' meaning that if your stream is replicated, all of the nats server holding a copy of the data will participate in servicing the client's get requests, not just the leader, which very much increases the scalability.

**Reply by jnmoyne** [reply]
Also note that the difference in latency between using storage memory or file for the KV (it's underlying stream) is that if the block where the message to get is already cached then file is almost as fast as memory (let's say around 50 micro-seconds order of magnitude) while if it is not then it needs to be first loaded from file into the cache which is obviously much higher latency (in the order of a couple milliseconds).

**Reply by magicprinc** [reply]
1) Thank you very much for such detailed answer! 🤝

2) The difference between FILE and MEMORY storage is surprisingly neglectable 🤷‍♀️
And besides, the dataset does not fit into memory.

3) Redis - surprisingly - showed a result comparable to NATS: The difference is 5-10% not times.

4) I use nats.java, so I have studied the inner workings of the KeyValue abstraction in nats.java and re-created my own with JetStream.publishAsync vs JetStream.publish
So I got 214_041 put op/s (vs 235_938 by RockDB → very similar numbers🚀)

But: I don't know how to speed up the KeyValue read operations using nats.java (no async or batch versions for current implementation)...
15k get/sec is not enough: I need > 100k get/sec

**Reply by jnmoyne** [reply]
> Redis - surprisingly - showed a result comparable to NATS: The difference is 5-10% not times.
Not surprising to me as it matches my own testing (actually I found that NATS could be a few percent faster than Redis on writes, and a few percent slower on reads (assuming the read is from memory or the block is already cached), but to the point that when you include the round trip over an actual network to the equation then the number are basically equivalent. So yeah, this may come as a surprise to some that make the wrong assumption that "Redis is the fastest KV around by a mile", but the facts show that NATS is basically just as fast as Redis (as long as the get operation hits the block file cache of the NATS server, or you use memory storage for the stream).

If only doing synchronous get operations one after the other in your client application code you are hitting the limit of sending the request to the nats-server process (over a TCP socket connection) and the result back (again over the TCP socket). Your limit is essentially due to the network round trip time between the client and server processes, not whether you are using NATS or Redis. The way to get higher throughput is as I mention before to use async operations, if your application all of a sudden needs to get the messages for keys A, B, and C you do 3 gets in 3 threads in parallel. Or you can rely on NATS's built in fast consumers meaning that rather than doing 3 KV get operations you do a KV Watch operation with 3 filters. Or you use the Stream API calls to create an ephemeral consumer (what the KV's watcher does underneath the covers) or using a 'multi-key get' operation on the stream.

**Reply by magicprinc** [reply]
If you think the network round trip time is the problem:
**Any plans to add KV getBatch(list of keys) ?**

As I said: MS(ANY)SQL in one-row mode is quite slow (let's say 1-3k rows/sec)
But! If you insert/update/delete in batches (1-5000 rows in a batch) in one transaction you can even touch 42k rows/sec and more..

**Reply by nickchomey** [reply]
There's a batched direct get mechanism in [ADR 31](https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-31.md#batched-requests). Not sure if it has been implemented yet or not - it doesn't seem to be documented (or at least not well documented). The team will obviously be able to confirm.

**Reply by magicprinc** [reply]
Same NATS KV "get", but in 10 Threads, 5 000 000 random key reads total:
total speed op/s=150_566 🤝👍

**Reply by nickchomey** [reply]
Could you please share some more on how you did that?

**Reply by magicprinc** [reply]
⌛ publishing quick and dirty benchmaks project

**Reply by magicprinc** [reply]
Benchmarks with sources and results
https://github.com/magicprinc/NatsDemo

NATS in Docker is very slow 😱
NATS without Docker (running on host OS) is fast 👍

**Reply by jnmoyne** [reply]
Obviously since the limiting factor is the latency of sync operations (mostly the latency of going back and forth over a socket to talk to the nats-server process) you get higher throughput numbers by either using asynchronous operations or multiple threads: i.e. you should get similar throughput whether doing 10 async get operations at a time, or having 10 threads doing sync get operations at the the same time (or doing a batched get with a size of 10 messages). 

Also: you can get very high throughput if you can leverage subject based addressing (since the KV's keys are actually (part of) a subject) and then use an 'ordered consumer' (what KV's `watch` uses under the covers) to get the messages.

**Reply by acormier-maia** [reply]
> Benchmarks with sources and results https://github.com/magicprinc/NatsDemo
> 
> NATS in Docker is very slow 😱 NATS without Docker (running on host OS) is fast 👍

I don't understand how there is a performance difference with docker vs bare-metal unless you were running in a VM (OSX)?
